src/firmware/app/control/trajectory_planner.c

```diff
@@ -228,155 +128,379 @@ app_trajectory_planner_interpolateConstantPeriodPositionTrajectory(
+            app_trajectory_planner_rebalanceAngularAndLinearTrajectorySegmentsForEquivalentDuration(
+                trajectory, segment_lengths, angular_segment_delta_time, i, false);
+            trajectory->trajectory_elements[i].time =
+                trajectory->trajectory_elements[i - 1].time + angular_segment_delta_time;
+        }
+        else if (linear_segment_delta_time > angular_segment_delta_time)
+        {
+            // In this case the angular velocity needs to be scaled down
+            app_trajectory_planner_rebalanceAngularAndLinearTrajectorySegmentsForEquivalentDuration(
+                trajectory, segment_lengths, linear_segment_delta_time, i, true);
+            trajectory->trajectory_elements[i].time =
+                trajectory->trajectory_elements[i - 1].time + linear_segment_delta_time;
+        }
+        else
+        {
+            // Catch-all case that there is no change in position for both the linear and
+            // angular paths
+            trajectory->trajectory_elements[i].time =
+                trajectory->trajectory_elements[i - 1].time + 0.0;
+        }
     }
 }
 
-void app_trajectory_planner_reversePositionTrajectoryDirection(
-    PositionTrajectory_t* forwards_trajectory)
+TrajectoryPlannerGenerationStatus_t
+app_trajectory_planner_rebalanceAngularAndLinearTrajectorySegmentsForEquivalentDuration(
+    PositionTrajectory_t* trajectory, TrajectorySegment_t* segment_lengths,
+    const float desired_segment_duration, unsigned int trajectory_index,
+    bool rebalance_angular)
 {
-    PositionTrajectoryElement_t reverse[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
-    float reverse_speed_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
-    const unsigned int num_segments = forwards_trajectory->path_parameters.num_segments;
+    // Initialize variables and values for the type of trajectory being modified
+    float current_speed  = 0;
+    float segment_length = 0;
 
-    const float path_duration =
-        forwards_trajectory->trajectory_elements[num_segments - 1].time;
+    if (rebalance_angular)
+    {
+        current_speed =
+            trajectory->trajectory_elements[trajectory_index - 1].angular_speed;
+        segment_length = segment_lengths[trajectory_index - 1].angular_segment_length;
+    }
+    else
+    {
+        current_speed =
+            trajectory->trajectory_elements[trajectory_index - 1].linear_speed;
+        segment_length = segment_lengths[trajectory_index - 1].linear_segment_length;
+    }
 
-    for (unsigned int i = 0; i < num_segments; i++)
+    // Calculate the new final velocity based on the initial velocity, segment length, and
+    // the desired segment duration
+    const float next_speed =
+        (2 * segment_length / desired_segment_duration) + current_speed;
+
+    if (rebalance_angular)
+    {
+        trajectory->trajectory_elements[trajectory_index].angular_speed = next_speed;
+    }
+    else
```

Comment:
I don't think it would be too much repeated code if we were to break this up into independent functions for linear and angular, what do you think?

Reply:
Replaced the bool parameter with an enum